src/librustc_trans/trans/adt.rs

```diff
@@ -438,23 +469,33 @@ struct Case<'tcx> {
 /// This represents the (GEP) indices to follow to get to the discriminant field
 pub type DiscrField = Vec<usize>;
 
-fn find_discr_field_candidate<'tcx>(tcx: &ty::ctxt<'tcx>,
+fn find_discr_field_candidate<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                                     ty: Ty<'tcx>,
-                                    mut path: DiscrField) -> Option<DiscrField> {
+                                    mut path: DiscrField) -> Option<(DiscrField, ValueRef)> {
+    let ref tcx = cx.tcx();
     match ty.sty {
-        // Fat &T/&mut T/Box<T> i.e. T is [T], str, or Trait
+        // Fat &T/&mut T/Box<T> i.e. T is [T], str, or Trait: null is forbidden
         ty::TyRef(_, ty::TypeAndMut { ty, .. }) | ty::TyBox(ty) if !type_is_sized(tcx, ty) => {
             path.push(FAT_PTR_ADDR);
-            Some(path)
+            Some((path, C_null(type_of::sizing_type_of(cx, ty))))
```

Comment:
The same issue about setting the forbidden value to the correct type I mentioned above. In this case, the `ty` used here is shadowing the passed in `ty` but both would be wrong :P The field we're using for the discriminant in this case is the first word of the fat pointer pair. But, the passed in `ty` is the whole fat pointer pair, while the `ty` in the match arm body is a `[T]`, `str` or `Trait` and thus we can't actually call `sizing_type_of` since they're unsized. 

Looking at the code again, `C_null(Type::i8p(cx))` should work here since a fat pointer pair is just that:  `Type::struct_(cx, &[Type::i8p(cx), Type::i8p(cx)], false)`